File: csharp/prime-factors/PrimeFactors.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class PrimeFactors
{
    public static long[] Factors(long number) => CalculateFactors(number).ToArray();

    private static IEnumerable<long> CalculateFactors(long number)
    {
        var divisor = 2;
        var left = number;

        while (left > 1)
        {
            if (left % divisor == 0)
            {
                left /= divisor;
                yield return divisor;
            }
            else
            {
                divisor++;
            }
        }
    }
}

File: kotlin/prime-factors/src/main/kotlin/PrimeFactors.kt
object PrimeFactorCalculator {

    fun primeFactors(int: Int): List<Int> =
        generatePrimeFactors(int.toLong())
            .map { it.toInt() }
            .toList()

    fun primeFactors(long: Long): List<Long> =
        generatePrimeFactors(long)
            .toList()

    private fun generatePrimeFactors(n: Long) = sequence {
        var divisor = 2L
        var left = n

        while (left > 1) {
            if (left % divisor == 0L) {
                left /= divisor
                yield(divisor)
            } else {
                divisor++
            }
        }
    }
}
